gc/structs/ForwardedHeader.hpp

```diff
@@ -263,14 +347,14 @@ class MM_ForwardedHeader
 	MMINLINE void
 	commenceFixup(omrobjectptr_t destinationObjectPtr)
 	{
-		MutableHeaderFields* newHeader = (MutableHeaderFields *)((fomrobject_t *)destinationObjectPtr + _forwardingSlotOffset);
-		
-		/* copy preserved flags, and keep the rest (which should be all 0s) */
-		newHeader->slot = (_preserved.slot & ~_copyProgressInfoMask) | _beingCopiedTag;
-
-#if defined (OMR_GC_COMPRESSED_POINTERS)
-		newHeader->overlap = _preserved.overlap;
-#endif /* defined (OMR_GC_COMPRESSED_POINTERS) */
+		uintptr_t mask = ~_copyProgressInfoMask;
+#if defined(OMR_GC_COMPRESSED_REFERENCES)
+		if (compressObjectReferences()) {
+			/* Do not remove the overlap when masking */
+			mask |= 0xFFFFFFFF00000000;
+		}
+#endif /* defined(OMR_GC_COMPRESSED_REFERENCES) */
+		*(uintptr_t*)destinationObjectPtr = flipValue(getCanonicalPreserved() & mask);
```

Comment:
A bit of info that may help you avoid 'is compressed' runtime check doing twice (explicit one for building proper mask and implicit one in flipValue): overlap slot does not have to be built this early. It can be done later, in commitFixup, but still prior to store fence. You can just build the class slot here, for what you dont need the runtime check (you'll end up with just one for overlap slot).

Reply:
... or I'm wrong, since getPreservedClassAndTags() also has a runtime check :(

Reply:
Are you saying the overlap does not need to be restored here? If so, I can use the writeClassSlot API.

Reply:
I believe so, not here, but then sometime later should (early in commitFixup). It's a bit of a difference in behavior, but still correct. 

I was originally concerned by extra runtime cost (2 CR checks here and one in commitFixup), but it should not be a big deal, after all. The commence/commit protocol is used only for largish objects (>1K). Still, if you can simplify it, go ahead.

Reply:
I've looked at various ways to make this better, but they end up with the same number of checks, and an extra 32-bit write (instead of one full word write here), so I'd be more comfortable not changing when the overlap is restored. I'll improve the comment.

Reply:
understood